pkg/cacheutil/memcached_client.go

```diff
@@ -0,0 +1,448 @@
+	// to the max concurrency we have.
+	c.workers.Add(c.config.MaxAsyncConcurrency)
+	for i := 0; i < c.config.MaxAsyncConcurrency; i++ {
+		go c.asyncQueueProcessLoop()
+	}
+
+	// Start a number of goroutines - processing get multi batch operations - equal
+	// to the max concurrency we have.
+	c.workers.Add(c.config.MaxGetMultiBatchConcurrency)
+	for i := 0; i < c.config.MaxGetMultiBatchConcurrency; i++ {
+		go c.getMultiQueueProcessLoop()
+	}
+
+	return c, nil
+}
+
+func (c *memcachedClient) Stop() {
+	close(c.stop)
+
+	// Wait until all workers have terminated.
+	c.workers.Wait()
+}
+
+func (c *memcachedClient) SetAsync(ctx context.Context, key string, value []byte, ttl time.Duration) error {
+	return c.enqueueAsync(func() {
+		start := time.Now()
+		c.operations.WithLabelValues(opSet).Inc()
+
+		span, _ := tracing.StartSpan(ctx, "memcached_set")
+		err := c.client.Set(&memcache.Item{
+			Key:        key,
+			Value:      value,
+			Expiration: int32(time.Now().Add(ttl).Unix()),
+		})
+		span.Finish()
+		if err != nil {
+			c.failures.WithLabelValues(opSet).Inc()
+			level.Warn(c.logger).Log("msg", fmt.Sprintf("failed to store item with key %s to memcached", key), "err", err)
+			return
+		}
+
+		c.duration.WithLabelValues(opSet).Observe(time.Since(start).Seconds())
+	})
+}
+
+func (c *memcachedClient) GetMulti(ctx context.Context, keys []string) map[string][]byte {
+	batches, err := c.getMultiBatched(ctx, keys)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to fetch keys from memcached", "err", err)
+
+		// In case we have both results and an error, it means some batch requests
+		// failed and other succeeded. In this case we prefer to log it and move on,
+		// given returning some results from the cache is better than returning
+		// nothing.
+		if len(batches) == 0 {
+			return nil
+		}
+	}
+
+	hits := map[string][]byte{}
+	for _, items := range batches {
+		for key, item := range items {
+			hits[key] = item.Value
+		}
+	}
+
+	return hits
+}
+
+func (c *memcachedClient) getMultiBatched(ctx context.Context, keys []string) ([]map[string]*memcache.Item, error) {
+	// Do not batch if the input keys are less then the max batch size.
+	if (c.config.MaxGetMultiBatchSize <= 0) || (len(keys) <= c.config.MaxGetMultiBatchSize) {
+		items, err := c.getMultiSingle(ctx, keys)
+		if err != nil {
+			return nil, err
+		}
+
+		return []map[string]*memcache.Item{items}, nil
+	}
+
+	// Calculate the number of expected results.
+	batchSize := c.config.MaxGetMultiBatchSize
+	numResults := len(keys) / batchSize
+	if len(keys)%batchSize != 0 {
+		numResults++
+	}
+
+	// Split input keys into batches and schedule a job for it.
+	results := make(chan *memcachedGetMultiResult, numResults)
+	defer close(results)
+
+	go func() {
+		for batchStart := 0; batchStart < len(keys); batchStart += batchSize {
+			batchEnd := batchStart + batchSize
+			if batchEnd > len(keys) {
+				batchEnd = len(keys)
+			}
+
+			c.getMultiQueue <- &memcachedGetMultiBatch{
```

Comment:
We lost the discussion on GitHub but I guess at the end we decided on per client concurrency for now? 

Thinking loudly: if we choose this I would say we need good observability: how to find latency caused by starvation of this queue? The latency metric is per single call to Memcache (which is fair for from mameched client perspective).

I think tracing would help as we already create some spans here e.g in `getMultiSingle` so should be enough..

Reply:
Maybe our current gate https://github.com/thanos-io/thanos/blob/master/pkg/store/gate.go could be adopted here somehow? It would be nice to have some metric here because otherwise all users would need to have some kind of tracing enabled.

Reply:
I like the metric standpoint, that's a good point. Do you think @pracucci we can adapt the existing way of queuing requests via `gate` in this way? as suggested by @GiedriusS ? I don't have a strong opinion but can see those pros & cons:

Advantages:
* Reuse
* Latency + queue size metric
* We can switch from global to per request limit easily `gate` here and there.

Downsides:
* Each gated request requires a new goroutine. Bit more goroutine overhead and coordination.

I think for this iteration I would be happy with what we have now.

Reply:
I personally see a big value in having such metrics. Also the gate would simplify the implementation (less code).

> We lost the discussion on GitHub but I guess at the end we decided on per client concurrency for now?

Re-iterating on this, I think the current implementation is confusing. We're currently limiting the batched `GetMulti()`  but not non-batched `GetMulti()`. For example, if you have `MaxGetMultiBatchSize=1000` and `MaxAsyncConcurrency=10` and you concurrently call 100 times `MemcachedClient.GetMulti(1K keys)` we end up with 100 concurrent underlying requests (each with 1K keys for a total of 100K keys). However, if you call once `MemcachedClient.GetMulti(100K keys)` we end up with 10 concurrent underlying requests, each with 1K keys, until all 100K keys are fetched.

I see three options:
1. Keep it as is (but weird)
2. Enforce the concurrency limit even to non-batched requests
  - Pro: the concurrency limit is an effective max number of concurrent fetch requests
  - Con: a single request with a bunch of keys will slow down everything else because of the queue
3. Switch to a per-request gate
  - Pro: a single request with a bunch of keys doesn't slow down others
  - Con: there's no way to set a cap on the max number of connections towards memcached

I'm currently more keen to #2, cause it looks the safest option.

Reply:
> Do you think @pracucci we can adapt the existing way of queuing requests via gate in this way? as suggested by @GiedriusS ?

I think so and I believe it would make sense. Observability - at this stage - is more important than goroutine overhead and we could implement it in a way that goroutines are created only for batched requests (so as far as the batching is disabled or requests do not exceed the max batch size, there will be no goroutine overhead).

Reply:
In [this commit](https://github.com/thanos-io/thanos/pull/1881/commits/ee88dd5c269ab5caf76da449dc42399a67c8e10c) I've:
1. Switched the max get multi concurrency to a gate
2. Applied the gate also to non batched requests

What's your take?